### ISSUE TITLE
fix(modelengine): refuse generic prefix hits that cannot restore KV state

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12606]
# fix(modelengine): refuse generic prefix hits that cannot restore KV state

```routing
lane: modelengine
paths: ["internal/modelengine/nativesched_prefill.go", "internal/modelengine/nativesched_prefill_test.go"]
expected_steps: 4
```

Audited snapshot: `6872f998b465d96633c3356ef2382c97fc98cc7c`.

## Parent context

Correctness follow-up to #12530 for the generic `SetPrefixTree` integration seam.

## Problem frame

- Centrality: Core
- P1 Context: preserved - a cache hit may skip prefill only when the matching model state is restored into the request-owned session.
- P2 Net value: advanced - prevents cached requests from silently decoding from an empty KV cache.
- P3 Adaptation: preserved - RadixKV hits with a concrete state-bearing boundary remain eligible; generic match-only indexes fail closed to ordinary prefill.
- P4 Operations: advanced - the regression checks live cache length at the first decode boundary instead of trusting hit counters.

## Current state

`NativeScheduler.SetPrefixTree` accepts a generic `PrefixTree`, whose only contract is `MatchLen([]int) int`. For a full match, `lookupPrefix` returns the prompt length and a nil RadixKV boundary (`nativesched_prefill.go:201-215`). `qwenPrefillChunkBudget` records that as a hit (`:325-341`), and `advanceQwenPrefill` marks the lane decode-ready without prefilling the live session (`:493-530`). It computes first-token logits in a temporary session (`:508-513`) but discards that session and its KV.

The next decode therefore advances a live cache containing only the first emitted token. The deterministic witness observes `Cache.Len() == 1` where a 48-token prompt plus one emitted token requires `49`.

This is reachable whenever a native scheduler enables chunked Qwen prefill and installs a generic match-only prefix index through `SetPrefixTree`. The current repository's product configuration uses the stronger `SetRadixKV` path; the generic integration seam is callable but has no in-tree production installer today.

## Why this is next

The public scheduler API currently accepts this configuration and silently produces incomplete request-owned KV state. Closing the state-authority gap before another installer adopts the generic seam avoids turning an exposed latent defect into a serving regression.

## Core through-line

Only state-bearing prefix hits may shorten prefill. Keep generic `PrefixTree` useful for observation or scheduling hints, but do not move `promptCursor`, `promptLen`, or lane state past tokens whose KV state was not restored into `ln.sess`.

## Working spine

Prompt -> match-only lookup -> state-authority check -> ordinary bounded prefill when state is absent -> decode from a request-owned full KV cache.

## Gold-plating boundary

Do not redesign RadixKV, add a new cache hierarchy, change token selection, or optimize the cold prefill path. This leaf only makes the generic match-only path state-safe.

## Suggested implementation steps

1. Separate match-only lookup telemetry from state-bearing cache admission.
2. Require a boundary with restorable KV before advancing the prompt cursor or transitioning directly to decode.
3. Route generic match-only hits through bounded ordinary prefill while preserving their observation counters.
4. Add the regression below and retain the existing RadixKV full/partial-hit witnesses.

## Done condition

- [ ] A generic `PrefixTree` full match cannot transition directly to decode without restoring equivalent live KV state.
- [ ] The request-owned cache contains the full prompt before the first generated token is advanced.
- [ ] Existing state-bearing RadixKV full-hit and partial-hit tests remain green.
- [ ] The focused regression passes without hardware and without a skipped test.

## Definition of done

The standalone regression and the existing RadixKV prefix-aware prefill tests pass, and no match-only cache index can advance decode state without request-owned KV.

## Acceptance gate

The focused regression and the existing native scheduler prefix tests pass, with the regression observing `Cache.Len() == len(prompt)+1` after the first emitted token.

## Closure binding

The resolving commit cites this issue and carries `(fak modelengine)`.

## Witness

Add this standalone regression to `internal/modelengine/nativesched_prefill_test.go`:

```go
type auditFullMatchPrefixTree struct{}

func (auditFullMatchPrefixTree) MatchLen(tokens []int) int { return len(tokens) }

func TestGenericPrefixTreeFullHitCannotSkipKVPrefill(t *testing.T) {
	const budget = nativeQwenPrefillMinChunkTokens
	m := nativeSchedulerPrefillModel(t)
	prompt := nativeSchedulerQwenPrompt(48)
	prepare := nativeSchedulerPrefillPrepare(map[string][]int{"full-hit": prompt})
	s := newNativeScheduler(m, prepare)
	if err := s.SetQwenPrefillMaxTokensPerIteration(budget); err != nil {
		t.Fatalf("SetQwenPrefillMaxTokensPerIteration: %v", err)
	}
	s.SetPrefixTree(auditFullMatchPrefixTree{})
	nativeSchedulerBeginManualDrain(t, s)

	req := nativeSchedulerAdmitLane(t, s, "full-hit")
	var got []int
	for i := 0; i < 4 && len(got) == 0; i++ {
		nativeSchedulerDriveIteration(t, s)
		got = nativeSchedulerDrainAvailable(req)
	}
	if len(got) != 1 {
		t.Fatalf("first decode tokens after bounded prefill = %v, want one", got)
	}

	s.mu.Lock()
	cacheLen := req.sess.Cache.Len()
	s.mu.Unlock()
	if want := len(prompt) + 1; cacheLen != want {
		t.Fatalf("live session KV length after full-hit decode = %d, want %d (prompt plus emitted token)", cacheLen, want)
	}

	req.Cancel()
	nativeSchedulerEndManualDrain(s)
}
```

Expected fixed-tree command:

```bash
go test ./internal/modelengine -run '^TestGenericPrefixTreeFullHitCannotSkipKVPrefill$' -count=1 -v
```

## Verifiable Witness

The audit ran the standalone test through `go test -overlay` without modifying the source tree. After adding the test under its final name, rerun:

```bash
go test ./internal/modelengine -run '^TestGenericPrefixTreeFullHitCannotSkipKVPrefill$' -count=1 -v
```

Captured failure:

```text
=== RUN   TestAuditGenericPrefixTreeFullHitCannotSkipKVPrefill
    audit_native_scheduler_generic_prefix_test.go:35: live session KV length after full-hit decode = 1, want 49 (prompt plus emitted token)
--- FAIL: TestAuditGenericPrefixTreeFullHitCannotSkipKVPrefill (0.02s)
FAIL github.com/anthony-chaudhary/fak/internal/modelengine 0.044s
```

## Likely files

- `internal/modelengine/nativesched_prefill.go`
- `internal/modelengine/nativesched_prefill_test.go`

Exact seams at the audited SHA:

- `internal/modelengine/nativesched_prefill.go:145-154` (`SetPrefixTree`)
- `internal/modelengine/nativesched_prefill.go:178-218` (`lookupPrefix`)
- `internal/modelengine/nativesched_prefill.go:306-341` (`qwenPrefillChunkBudget`)
- `internal/modelengine/nativesched_prefill.go:481-530` (`advanceQwenPrefill`)
- `internal/modelengine/nativesched_prefill_test.go`

## Lane

modelengine

## Expected steps

4

## Blast radius / fallback

Affected: native chunked Qwen prefill configured with a generic match-only `PrefixTree`; token one uses temporary-session logits, while subsequent decode state is based on incomplete KV. Unaffected: cold prefill and RadixKV boundaries carrying cloned KV and cached logits. The safe fallback is to treat match-only results as cold for state advancement. No hardware claim is involved.

## Dedupe

This is a correctness follow-up to open #12530, which introduced prefix-aware scheduling and owns the same two files. #12530's existing witness covers RadixKV state-bearing boundaries; it does not cover `SetPrefixTree` with a nil boundary. Exact-symbol issue searches for `SetPrefixTree` and `generic PrefixTree` found no issue owning this failure.
